standard/array.lua

```diff
@@ -0,0 +1,337 @@
+
+The optional third argument specifies a custom comparator for the transformed elements.
+
+Examples:
+Array.sortBy({-3, -1, 2, 4}, function(x) return x * x end)
+-- returns {-1, 2, -3, 4}
+
+Array.sortBy({
+	{first='Neil', last='Armstrong'},
+	{first='Louis', last='Armstrong'},
+	{first='Buzz', last='Aldrin'},
+}, function(x) return {x.last, x.first} end)
+-- returns {
+--	{first='Buzz', last='Aldrin'},
+--	{first='Louis', last='Armstrong'},
+--	{first='Neil', last='Armstrong'},
+-- }
+
+]]
+function Array.sortBy(tbl, f, compare)
+	local tbl2 = Table.copy(tbl)
+	Array.sortInPlaceBy(tbl2, f, compare)
+	return tbl2
+end
+
+--[[
+Like Array.sortBy, except that it sorts in place. Mutates the first argument.
+]]
+function Array.sortInPlaceBy(tbl, f, compare)
+	compare = compare or function(y1, y2)
+		if type(y1) == 'table' and type(y2) == 'table' then
+			return Array.lexicalCompare(y1, y2)
+		else
+			return y1 < y2
+		end
+	end
+
+	table.sort(tbl, function(x1, x2) return compare(f(x1), f(x2)) end)
+end
+
+-- Reverses the order of elements in an array.
+function Array.reverse(tbl)
+	local out_tbl = {}
+	for i = #tbl, 1, -1 do
+		table.insert(out_tbl, tbl[i])
+	end
+	return out_tbl
+end
+
+--[[
+Returns an array with elements append to the end. Does not mutate the inputs.
+
+Example:
+Array.append({2, 3}, 5, 7, 11)
+-- returns {2, 3, 5, 7, 11}
+]]
+function Array.append(tbl, ...)
+	return Array.appendWith(Array.copy(tbl), ...)
+end
+
+--[[
+Adds elements to the end of an array. The array is mutated in the process.
+]]
+function Array.appendWith(tbl, ...)
+	local elems = Table.pack(...)
+	for i = 1, elems.n do
+		if elems[i] ~= nil then
+			table.insert(tbl, elems[i])
+		end
+	end
+	return tbl
+end
+
+--[[
+Returns an array with elements from one or more arrays append to the end. Does
+not mutate the inputs.
+
+Example:
+Array.extend({2, 3}, {5, 7, 11}, {13})
+-- returns {2, 3, 5, 7, 11, 13}
+
+Array.extend({2, 3}, 5, 7, nil, {11, 13})
+-- returns {2, 3, 5, 7, 11, 13}
+]]
+function Array.extend(tbl, ...)
+	return Array.extendWith({}, tbl, ...)
+end
+
+--[[
+Adds elements from one or more arrays to the end of a target array. The target
+array is mutated in the process.
+]]
+function Array.extendWith(tbl, ...)
+	local arrays = Table.pack(...)
+	for i = 1, arrays.n do
+		if type(arrays[i]) == 'table' then
+			for _, elem in ipairs(arrays[i]) do
+				table.insert(tbl, elem)
+			end
+		elseif arrays[i] ~= nil then
+			table.insert(tbl, arrays[i])
+		end
+	end
+	return tbl
+end
+
+--[[
+Returns the array {f(1), f(2), f(3), ...}. Stops before the first nil value returned by f.
+
+Example:
+Array.mapIndexes(function(x) return x < 5 and x * x or nil end)
+-- returns {1, 4, 9, 16}
+]]
+function Array.mapIndexes(f)
+	local arr = {}
+	for i = 1, math.huge do
+		local y = f(i)
+		if y then
+			table.insert(arr, y)
+		else
+			break
+		end
+	end
+	return arr
+end
+
+function Array.fromTableKeys(tbl)
```

Comment:
This name doesn't really make sense. Looks like it extracts keys?

```suggestion
function Array.extractKeys(tbl)
```

Reply:
since it is a copy from live i doubt that changing existing function names is a good idea
we would have to add a "redirect" for them ...

Reply:
I added this recently, so there's not many call sites yet. Outside of this repo I think it's just `Module:Performance/Util`.

Reply:
@warnull could you fix this in a follow-up PR?

Reply:
#549